File: packages/ondex_core/lib/src/restaurant_status.dart
/// Restoranning "Ochiq/Yopiq" holati — barcha Flutter ilovalari uchun
/// YAGONA o'quvchi.
///
/// ┌─ NEGA (2026-09-15) ───────────────────────────────────────────────┐
/// Ilovalar `restaurant['open']` ga qarardi — bu restoranning QO'LDA
/// bosadigan tugmasi. Ish vaqti tugagan restoran hamma joyda "Ochiq"
/// ko'rinib, mijoz taom qo'shgach faqat savatda "yopiq" javobini olardi,
/// restoran panelidagi belgi ham "Ochiq" deb turardi.
///
/// Endi holatni SERVER hisoblaydi (`open_now`, `closed_reason`,
/// `open_changes_at` — `internal/catalog/hours.go`), bu yer faqat o'qiydi
/// va matnga aylantiradi. Ish vaqti qoidasi bu yerda QAYTA YOZILMAYDI.
/// Web'da aynan shu vazifani `apps/web/lib/restaurant-status.ts` bajaradi.
/// └───────────────────────────────────────────────────────────────────┘
library;

enum RestaurantClosedReason { manual, hours }

/// Ish vaqti Toshkent vaqtida (server bilan bir xil qat'iy +05:00).
const _tashkentOffset = Duration(hours: 5);

const _weekdays = [
  'Dushanba',
  'Seshanba',
  'Chorshanba',
  'Payshanba',
  'Juma',
  'Shanba',
  'Yakshanba',
];

class RestaurantOpenStatus {
  const RestaurantOpenStatus({required this.open, this.reason, this.changesAt});

  /// Hozir buyurtma qabul qiladi.
  final bool open;

  /// Yopiq bo'lsa sababi, ochiq bo'lsa `null`.
  final RestaurantClosedReason? reason;

  /// Holat o'z-o'zidan o'zgaradigan payt (UTC), server hisoblagan.
  final DateTime? changesAt;

  /// "Yopiladi" izohi yopilishga shuncha qolganda ko'rsatiladi.
  static const closingSoon = Duration(hours: 1);

  /// Restoran javobidan (`GET /restaurants`, `GET /restaurants/{id}`).
  factory RestaurantOpenStatus.fromJson(Map<String, dynamic> r) {
    final manualOpen = r['open'] == true;
    final openNow = r['open_now'];
    // `open_now` bo'lmagan eski javob — avvalgi xatti-harakat (qo'lda tugma).
    final open = openNow is bool ? openNow : manualOpen;
    RestaurantClosedReason? reason;
    if (!open) {
      reason = switch (r['closed_reason']) {
        'hours' => RestaurantClosedReason.hours,
        'manual' => RestaurantClosedReason.manual,
        _ => manualOpen ? RestaurantClosedReason.hours : RestaurantClosedReason.manual,
      };
    }
    final raw = r['open_changes_at'];
    return RestaurantOpenStatus(
      open: open,
      reason: reason,
      changesAt: raw is String ? DateTime.tryParse(raw)?.toUtc() : null,
    );
  }

  /// Ekran ochiq turgan paytda [changesAt] o'tib ketsa — holat jadval
  /// bo'yicha almashadi. Qo'lda yopilgan restoran vaqt bilan ochilmaydi.
  RestaurantOpenStatus at(DateTime now) {
    final c = changesAt;
    if (c == null || reason == RestaurantClosedReason.manual || now.toUtc().isBefore(c)) {
      return this;
    }
    return open
        ? const RestaurantOpenStatus(open: false, reason: RestaurantClosedReason.hours)
        : const RestaurantOpenStatus(open: true);
  }

  String get label => open ? 'Ochiq' : 'Yopiq';

  /// Qisqa izoh: "09:00 da ochiladi", "Ertaga 09:00 da ochiladi",
  /// "Dushanba 09:00 da ochiladi", yopilishga bir soatdan kam qolganda
  /// "22:00 da yopiladi". Aytadigan narsa bo'lmasa `null`.
  String? detail(DateTime now) {
    final c = changesAt;
    if (c == null) return null;
    final at = c.toUtc().add(_tashkentOffset);
    final clock = '${_two(at.hour)}:${_two(at.minute)}';
    if (open) {
      return c.difference(now.toUtc()) <= closingSoon ? '$clock da yopiladi' : null;
    }
    final today = now.toUtc().add(_tashkentOffset);
    final days = DateTime.utc(at.year, at.month, at.day)
        .difference(DateTime.utc(today.year, today.month, today.day))
        .inDays;
    if (days <= 0) return '$clock da ochiladi';
    if (days == 1) return 'Ertaga $clock da ochiladi';
    return '${_weekdays[at.weekday - 1]} $clock da ochiladi';
  }

  /// Menyu/panel ogohlantirishi uchun to'liq jumla (yopiq holatda).
  String closedMessage(DateTime now) {
    final d = detail(now);
    if (d != null) return 'Restoran hozir yopiq · $d';
    return reason == RestaurantClosedReason.manual
        ? 'Restoran hozir buyurtma qabul qilmayapti'
        : 'Restoran hozir yopiq — ish vaqti tugagan';
  }
}

String _two(int v) => v.toString().padLeft(2, '0');

File: apps/web/lib/restaurant-status.ts
import { tashkentClock, tashkentDayNumber, tashkentWeekday } from "./tashkent-time";
import type { Restaurant } from "./types";

// Restoranning "Ochiq/Yopiq" holati — web'dagi YAGONA manba.
//
// ┌─ NEGA (2026-09-15) ────────────────────────────────────────────────┐
// Kartalar va menyu `restaurant.open` ga qarardi — bu restoranning qo'lda
// bosadigan tugmasi, xolos. Ish vaqti tugagan restoran hamma joyda
// "Ochiq" ko'rinib, mijoz taom qo'shgach faqat savatda "yopiq" javobini
// olardi. Endi holatni SERVER hisoblaydi (`open_now`, `closed_reason`,
// `open_changes_at` — `internal/catalog/hours.go`); bu modul faqat o'qiydi
// va matnga aylantiradi. Ish vaqti qoidasi bu yerda QAYTA YOZILMAYDI.
//
// Flutter ilovalarida aynan shu vazifani
// `packages/ondex_core/lib/src/restaurant_status.dart` bajaradi.
// └────────────────────────────────────────────────────────────────────┘

export type ClosedReason = "manual" | "hours";

export type OpenStatus = {
  open: boolean;
  /** Yopiq bo'lsa sababi, ochiq bo'lsa `null`. */
  reason: ClosedReason | null;
  /** Holat o'z-o'zidan o'zgaradigan payt (server hisoblagan). */
  changesAt: Date | null;
};

export type OpenStatusFields = Pick<
  Restaurant,
  "open" | "open_now" | "closed_reason" | "open_changes_at"
>;

/** "Yopiladi" izohi yopilishga shuncha qolganda ko'rsatiladi. */
export const CLOSING_SOON_MS = 60 * 60 * 1000;

export function restaurantOpenStatus(r: OpenStatusFields): OpenStatus {
  // `open_now` bo'lmagan eski javob — avvalgi xatti-harakat (qo'lda tugma).
  const open = typeof r.open_now === "boolean" ? r.open_now : Boolean(r.open);
  let reason: ClosedReason | null = null;
  if (!open) {
    if (r.closed_reason === "hours" || r.closed_reason === "manual") {
      reason = r.closed_reason;
    } else {
      reason = r.open ? "hours" : "manual";
    }
  }
  const at = r.open_changes_at ? new Date(r.open_changes_at) : null;
  return {
    open,
    reason,
    changesAt: at && !Number.isNaN(at.getTime()) ? at : null,
  };
}

/**
 * Sahifa ochiq turgan paytda `changesAt` o'tib ketsa — holat jadval
 * bo'yicha almashadi (server shu paytni aynan shu qoida bilan hisoblagan).
 * Keyingi o'zgarish vaqti noma'lum bo'lib qoladi. Qo'lda yopilgan
 * restoran vaqt bilan ochilmaydi.
 */
export function openStatusAt(s: OpenStatus, now: Date): OpenStatus {
  if (!s.changesAt || s.reason === "manual" || now < s.changesAt) return s;
  return s.open
    ? { open: false, reason: "hours", changesAt: null }
    : { open: true, reason: null, changesAt: null };
}

/**
 * Qisqa izoh: "09:00 da ochiladi", "Ertaga 09:00 da ochiladi",
 * "Dushanba 09:00 da ochiladi", yopilishga bir soatdan kam qolganda
 * "22:00 da yopiladi". Aytadigan narsa bo'lmasa `null`.
 */
export function openStatusDetail(s: OpenStatus, now: Date): string | null {
  const at = s.changesAt;
  if (!at) return null;
  const clock = tashkentClock(at);
  if (s.open) {
    return at.getTime() - now.getTime() <= CLOSING_SOON_MS
      ? `${clock} da yopiladi`
      : null;
  }
  const days = tashkentDayNumber(at) - tashkentDayNumber(now);
  if (days <= 0) return `${clock} da ochiladi`;
  if (days === 1) return `Ertaga ${clock} da ochiladi`;
  return `${tashkentWeekday(at)} ${clock} da ochiladi`.trim();
}

/** Menyu ogohlantirishi uchun to'liq jumla (faqat yopiq holatda ma'noli). */
export function closedMessage(s: OpenStatus, now: Date | null): string {
  const detail = now ? openStatusDetail(s, now) : null;
  if (detail) return `Restoran hozir yopiq · ${detail}`;
  return s.reason === "manual"
    ? "Restoran hozir buyurtma qabul qilmayapti"
    : "Restoran hozir yopiq — ish vaqti tugagan";
}
